### ISSUE TITLE
Support AOT-based autowiring for test class instances

### ISSUE THREAD

**Issue by snicoll** [spring-projects-experimental/spring-native#1169]
AOT-generated context don't use autowiring at all and it's been working fine for the main context. However, the TCF does rely on it to inject dependencies in the test class itself (See `DependencyInjectionTestExecutionListener`). 

We need to research a way where we'd either be in control of creating the instance (in which case we can generate the code for it like we do for other classes), or enable `AutowiredAnnotationBeanPostProcessor#postProcessProperties` only for the tests that match the context at hand.

The former would be better as it means we don't have a special case for tests.

**Comment by sbrannen** [comment]
The first approach to a solution for this issue has been pushed in https://github.com/sbrannen/spring-native/commit/72c7e9d91780ae14042e16db882b32f1a92accad and is open for discussion.

As mentioned in the commit message, this approach uses the (previously private) `InjectionPointsSupplier` from `spring-aot` to query the methods and fields that need to be autowired. The new `AotDependencyInjectionTestExecutionListener` then uses modified code copied from `AutowiredAnnotationBeanPostProcessor` to perform the actually injection.

**Comment by sbrannen** [comment]
The second approach to a solution for this issue has been pushed in https://github.com/sbrannen/spring-native/commit/bd9af80274bf0ad5e0c5248829c5a126c9715748 and is open for discussion.

This approach does not use `InjectionPointsSupplier` from `spring-aot`  but rather relies on the standard functionality of `AutowiredAnnotationBeanPostProcessor` to perform the actual injection (without registering `AutowiredAnnotationBeanPostProcessor` in the `ApplicationContext`).

**Comment by sbrannen** [comment]
@snicoll, @bclozel, and @sdeleuze, please let me know your opinion on which route we should go (approach # 1 or # 2 above).

Once we come to a consensus, I can create a PR for further review if desired.

**Comment by snicoll** [comment]
I am on PTO today but I've added a few comment on the second approach. The first approach is a bit similar to what we have, except that it narrows down when the AABP kicks in. I am not ruling it out entirely, but we might have to once we actually try running a test in a native image.

The second approach looks promising but is copying too much code compared to what we already have. `TestContextBoostrapInitializer` already has a mapping between tests and its `SmartContextLoader`. We could imagine an extension of that that would be able to initialize an instance using generated code, something like:


```java
public static Map<String, ThrowableConsumer<InstanceSupplierContext>> getInstanceInitializers() {
	Map<String,ThrowableConsumer<InstanceSupplierContext>> entries = new HashMap<>();
	entries.put("com.example.sampleactuator.repository.SpeakerRepositoryTests", (instanceContext) -> // regular bootstrap code);
	return entries;
}
```

For the regular bootstrap code bits, `InjectionPointWriter` is private. Maybe we could offer a way to make some part of it public.

**Comment by sbrannen** [comment]
> The first approach is a bit similar to what we have, except that it narrows down when the AABPP kicks in. 

Yes, it's similar, but it also does not actually register the AABPP. Thus no other beans are adversely affected by its use.

> I am not ruling it out entirely, but we might have to once we actually try running a test in a native image.

FWIW, the `JUnitPlatformFeature` in Native Build Tools automatically registers each test class for full reflection:

https://github.com/graalvm/native-build-tools/blob/15bea361faf74d2f36655aef688995be17e59918/common/junit-platform-native/src/main/java/org/graalvm/junit/platform/JUnitPlatformFeature.java#L146

So at least reflection should not be an issue when the Native Build Tools plugins are used.

**Comment by snicoll** [comment]
> Yes, it's similar, but it also does not actually register the AABPP. Thus no other beans are adversely affected by its use.

Isn't it the same thing as the sentence that you quoted?

> FWIW, the JUnitPlatformFeature in Native Build Tools automatically registers each test class for full reflection:

I am aware of that and that wasn't my concern. We've tried supporting AABPP with the main use case in the past and GraalVM had to harvest too much stuff to support this feature.

**Comment by sbrannen** [comment]
> > Yes, it's similar, but it also does not actually register the AABPP. Thus no other beans are adversely affected by its use.
> 
> Isn't it the same thing as the sentence that you quoted?

Not quite. I was pointing out that the AABPP is not registered in the `ApplicationContext`. You had previously mentioned to me in a different channel that you wanted to avoid that. So I was highlighting that this approach achieves that goal.


> We've tried supporting AABPP with the main use case in the past and GraalVM had to harvest too much stuff to support this feature.

OK. How do you want to proceed? Shall we try the AABPP approach first to see how it works out in a native image?

**Comment by snicoll** [comment]
> OK. How do you want to proceed? Shall we try the AABPP approach first to see how it works out in a native image?

Yes. you can try it out locally and we can see how far we get.

**Comment by sbrannen** [comment]
> Yes. you can try it out locally and we can see how far we get.

After rebasing on `main`, reverting 5c8f389229d848a082c80148ab358de84f9bafbc,  and running `mvn -Pnative clean test` for the `commandlinerunner`, I'm seeing the following now (which didn't happen yesterday, AFAIK).

```
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] <HOME>/source/spring-native/samples/commandlinerunner/target/generated-test-sources/spring-aot/src/test/java/org/springframework/aot/StaticSpringFactories.java:[53,119] cannot find symbol
  symbol:   class AotDependencyInjectionTestExecutionListener
  location: package org.springframework.aot.test
```

Here's the problematic source code:

```java
factories.add(org.springframework.test.context.TestExecutionListener.class,
    () -> new org.springframework.aot.test.AotDependencyInjectionTestExecutionListener());
```

Is this a known issue that will be addressed elsewhere, or do I need to open a new ticket to track this?

**Comment by snicoll** [comment]
> Is this a known issue that will be addressed elsewhere, or do I need to open a new ticket to track this?

As we've discussed `AotDependencyInjectionTestExecutionListener` is code that you've added and `main` works so please investigate further. Let me know if you need help.

**Comment by sbrannen** [comment]
My apologies: I had mixed up when I was running the native tests for the command line runner.

The AABPP approach always worked with AOT/JVM but never worked with AOT/Native.

To resolve the issue, I moved the new TestExecutionListener to the `spring-native` module, and the command line runner tests now succeed with AOT mode while running within a native image.

The native testing results for the [status quo](https://github.com/spring-projects-experimental/spring-native/compare/main...sbrannen:issues/gh-1169-aot-test-dependency-injection-AutowiredAnnotationBeanPostProcessor) are as follows.

```
[native-tests:43681]    classlist:   2,235.58 ms,  0.94 GB
[native-tests:43681]        (cap):   1,196.85 ms,  1.18 GB
[native-tests:43681]        setup:   3,009.48 ms,  1.18 GB
[native-tests:43681]     (clinit):     447.84 ms,  4.13 GB
[native-tests:43681]   (typeflow):   3,955.62 ms,  4.13 GB
[native-tests:43681]    (objects):  15,075.55 ms,  4.13 GB
[native-tests:43681]   (features):   6,077.70 ms,  4.13 GB
[native-tests:43681]     analysis:  26,646.08 ms,  4.13 GB
[native-tests:43681]     universe:   1,735.65 ms,  4.13 GB
[native-tests:43681]      (parse):   1,321.02 ms,  4.13 GB
[native-tests:43681]     (inline):   3,187.07 ms,  4.47 GB
[native-tests:43681]    (compile):  12,304.54 ms,  5.13 GB
[native-tests:43681]      compile:  18,830.04 ms,  5.13 GB
[native-tests:43681]        image:   2,926.13 ms,  5.13 GB
[native-tests:43681]        write:     770.40 ms,  5.13 GB
[native-tests:43681]      [total]:  56,489.07 ms,  5.13 GB

JUnit Platform on Native Image - report
----------------------------------------

com.example.commandlinerunner.CommandlinerunnerApplicationTests > contextLoads() SUCCESSFUL


Test run finished after 15 ms
[         2 containers found      ]
[         0 containers skipped    ]
[         2 containers started    ]
[         0 containers aborted    ]
[         2 containers successful ]
[         0 containers failed     ]
[         1 tests found           ]
[         0 tests skipped         ]
[         1 tests started         ]
[         0 tests aborted         ]
[         1 tests successful      ]
[         0 tests failed          ]

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:09 min
[INFO] Finished at: 2021-11-03T12:27:10+01:00
[INFO] ------------------------------------------------------------------------
```

----

In summary, the **second approach** works (using the `AutowiredAnnotationBeanPostProcessor`), but the **first approach** (using the `BeanInstanceDescriptor`) will not work unless we switch to full AOT mode with a generated utility that performs the test instance autowiring (see https://github.com/spring-projects-experimental/spring-native/issues/1169#issuecomment-956026717 for details).

**Comment by snicoll** [comment]
If I understood correctly, we are processing `spring.factories` to make sure the listeners run in a native image. We could use the opportunity to swap `DependencyInjectionTestExecutionListener` by the AOT variant. This would require us to use a magic ordering number and would be more consistent with what we're doing elsewhere.

**Comment by sdeleuze** [comment]
@sbrannen Could you please provide an updated status for this issue?

**Comment by sbrannen** [comment]
In 7c600f721f44b681249502fc2af39811f6f7b5e5, I pushed the **second approach** using the AABPP in a new TestExecutionListener.

All of the key samples are working with `@Autowired` fields in test classes.

So we can go with this approach for now and revisit if we want to switch to the **first approach** later. 

In light of that, I am leaving this issue open for further discussion.

**Comment by sbrannen** [comment]
> If I understood correctly, we are processing `spring.factories` to make sure the listeners run in a native image.

That's correct.

> We could use the opportunity to swap `DependencyInjectionTestExecutionListener` by the AOT variant. This would [not] require us to use a magic ordering number and would be more consistent with what we're doing elsewhere.

Funny you should suggest that.

"Funny" because you suggested we do something like several weeks ago, and I told you it was not possible with the standard feature set of the TCF.

But... it turns out that Spring Boot 1.4.1 actually introduced a mechanism for doing just that, the `DefaultTestExecutionListenersPostProcessor` SPI supported by the `SpringBootTestContextBootstrapper` which uses `SpringFactoriesLoader` to find implementations.

There's currently one that I'm aware of: `org.springframework.boot.test.autoconfigure.SpringBootDependencyInjectionTestExecutionListener.PostProcessor`.

So now we have an interesting situation.

1. That `PostProcessor` in Boot is replacing the standard `DependencyInjectionTestExecutionListener` with `SpringBootDependencyInjectionTestExecutionListener`.
2. In AOT mode, the standard `DependencyInjectionTestExecutionListener` is only initializing the test instance (e.g, for *Aware interfaces, etc.).
3. In AOT mode (as of my last commit), the new `AotDependencyInjectionTestExecutionListener` is performing DI for `@Autowired` candidates in the test instance.

So, if an exception is thrown when `AotDependencyInjectionTestExecutionListener` does its work, the `SpringBootDependencyInjectionTestExecutionListener` will NOT print the `ConditionEvaluationReport`, meaning the extra diagnostics support from Spring Boot will be lost when tests execute in a native image and the application context fails to load (or autowiring fails).

> We could use the opportunity to swap `DependencyInjectionTestExecutionListener` by the AOT variant.

If we do that without writing a substitution for `SpringBootDependencyInjectionTestExecutionListener.PostProcessor`, that post processor will turn into a no-op since the `DependencyInjectionTestExecutionListener` will no longer be in the list of default listeners. That means the `SpringBootDependencyInjectionTestExecutionListener` won't be registered.

In any case, `SpringBootDependencyInjectionTestExecutionListener` is currently not invoking its `outputConditionEvaluationReport(..)` method for any errors that occur in the `AotDependencyInjectionTestExecutionListener`.

In summary, I think we need to decide how we want to proceed in terms of loss of `SpringBootDependencyInjectionTestExecutionListener` support.

**Comment by snicoll** [comment]
> "Funny" because you suggested we do something like several weeks ago, and I told you it was not possible with the standard feature set of the TCF.

That's not what I am talking about though. If we rewrite spring.factories, we could filter the entries as well. We're doing that in other places.

**Comment by sbrannen** [comment]
> That's not what I am talking about though. If we rewrite spring.factories, we could filter the entries as well. We're doing that in other places.

Yes, I know what you're talking about.

Spring Boot's mechanism kicks in after the list of class names for TestExecutionListeners has been loaded from `spring.factories`.

That post processor from Boot replaces `DependencyInjectionTestExecutionListener.class` with `SpringBootDependencyInjectionTestExecutionListener.class` in the list of class names to instantiate.

All of our sample applications currently run with the `SpringBootDependencyInjectionTestExecutionListener` instead of the `DependencyInjectionTestExecutionListener`.

If we rewrite `spring.factories`, `SpringBootDependencyInjectionTestExecutionListener` will not be used.

That was one of my points I was trying to make.

The other point I was trying to make is that `SpringBootDependencyInjectionTestExecutionListener` currently does not get applied to any exceptions thrown in the `AotDependencyInjectionTestExecutionListener` (or whatever mechanism we go with in the end).

**Comment by sdeleuze** [comment]
@snicoll @sbrannen Could you please synchronize to define remaining tasks required before closing this issue and update it accordingly?

**Comment by sbrannen** [comment]
Resolution of this issue is closely related to #1258.

**Comment by snicoll** [comment]
IMO one of those issues should be closed. It's quite confusing that a proposal for this issue has been pushed and a bug, in a separate issue has been raised while this one is still open.

**Comment by sdeleuze** [comment]
@sbrannen Since we have autowiring working on a sample and indeed #1258 opened, let's maybe close that one otherwise that's confusing (at least to me). We can create follow-up more focused issues for refinements.

**Comment by snicoll** [comment]
Closed by 7c600f721f44b6812. Let's follow-up on #1258.
